src/toolbox/shamir.rs

```diff
@@ -116,48 +94,70 @@ fn evaluate_lagrange(x: Scalar, xis: &Vec<Scalar>, yis: &Vec<Scalar>) -> Result<
     Ok(sum)
 }
 
-impl SecretShare {
+impl<R> Shamir<R> where R: RngCore + CryptoRng {
+    pub fn new(threshold: usize, rng: R) -> Self {
+        Shamir {
+            threshold,
+            rng
+        }
+    }
+}
+
+impl Shamir<ThreadRng> {
+    pub fn new_without_rng(threshold: usize) -> Self {
+        Shamir { threshold , rng: rand::thread_rng() }
+    }
+}
+
+impl<R> secrets::SecretSharing for Shamir<R> where R: RngCore + CryptoRng {
 
     /// Given a secret and some other information, calculate a set of shares using Shamir's Secret Sharing.  We don't
     /// expect this code to ever be called by our application, but it's useful for testing.
-    pub fn share<R: RngCore + CryptoRng>(secret: &Scalar, nr_of_shares: usize, threshold: usize, rng: &mut R) -> SecretShare {
+    fn share(&mut self, secret: &Scalar, nr_of_shares: usize) -> Vec<Option<Scalar>> {
 
         // first, select random coefficients for each term (less the first) in the polynomial
         let mut coefficients: Vec<Scalar> = Vec::new();
         let my_secret = secret.clone();
         coefficients.push(my_secret);  // the coefficient for x^0 is the secret
-        for _ in 1..(threshold) {
-            coefficients.push(Scalar::random(rng));
+        for _ in 1..(self.threshold) {
+            coefficients.push(Scalar::random(&mut self.rng));
         }
 
         // now, calculate points on the line as our shares.
         // we encode the x value as the index of the vec
-        let mut shares: Vec<Scalar> = Vec::new();
+        let mut shares: Vec<Option<Scalar>> = Vec::new();
         for i in 0..nr_of_shares {
             let x = Scalar::from((i+1) as u64);
             let y = calc_polynomial(&coefficients, &x);
-            shares.push(y);
+            shares.push(Some(y));
```

Comment:
Here it shows up, there isn't gonna be any `None` so should output `Vec<Scalar>` in my opinion